### ISSUE TITLE
Maybe a bug about Hessian matrix of regression_l1 objective function

### ISSUE THREAD

**Issue by WXFMAV** [Microsoft/LightGBM#979]
## Maybe a bug  about Hessian matrix of regression_l1 objective function

There may be a bug in Hessian matrix of regression_l1 objective function.

It was strange and unreasonable in function **ApproximateHessianWithGaussian** that the gaussian variance is set as 
**const double c = std::max((std::fabs(y) + std::fabs(t)) * eta, 1.0e-10);** 

The proper code maybe as 
**const double c = std::max((std::fabs(y-t)) * eta, 1.0e-10);** 

Though I am not very sure if this is a bug, my program fails in weighted sampling as is reported in issue https://github.com/Microsoft/LightGBM/issues/939

**reference:**
The Code in File: https://github.com/Microsoft/LightGBM/blob/master/src/objective/regression_objective.hpp
between line 98 and line 123
```
  void GetGradients(const double* score, score_t* gradients,
                    score_t* hessians) const override {
    if (weights_ == nullptr) {
      #pragma omp parallel for schedule(static)
      for (data_size_t i = 0; i < num_data_; ++i) {
        const double diff = score[i] - label_[i];
        if (diff >= 0.0f) {
          gradients[i] = 1.0f;
        } else {
          gradients[i] = -1.0f;
        }
        hessians[i] = static_cast<score_t>(Common::ApproximateHessianWithGaussian(score[i], label_[i], gradients[i], eta_));
      }
    } else {
      #pragma omp parallel for schedule(static)
      for (data_size_t i = 0; i < num_data_; ++i) {
        const double diff = score[i] - label_[i];
        if (diff >= 0.0f) {
          gradients[i] = weights_[i];
        } else {
          gradients[i] = -weights_[i];
        }
        hessians[i] = static_cast<score_t>(Common::ApproximateHessianWithGaussian(score[i], label_[i], gradients[i], eta_, weights_[i]));
      }
    }
  }

```

And the code in File:
https://github.com/Microsoft/LightGBM/blob/87fa8b542e0027e03027197ba07aab2812d6755d/include/LightGBM/utils/common.h
between line 486 and line 495
```
inline static double ApproximateHessianWithGaussian(const double y, const double t, const double g,
                                                    const double eta, const double w=1.0f) {
  const double diff = y - t;
  const double pi = 4.0 * std::atan(1.0);
  const double x = std::fabs(diff);
  const double a = 2.0 * std::fabs(g) * w;  // difference of two first derivatives, (zero to inf) and (zero to -inf).
  const double b = 0.0;
  const double c = std::max((std::fabs(y) + std::fabs(t)) * eta, 1.0e-10);
  return w * std::exp(-(x - b) * (x - b) / (2.0 * c * c)) * a / (c * std::sqrt(2 * pi));
}

```

**Comment by Laurae2** [comment]
ping @henry0312

**Comment by Tony-Y** [comment]
@WXFMAV  Please see [Representations of the delta function](https://en.wikipedia.org/wiki/Dirac_delta_function#Representations_of_the_delta_function).
This code employs the heat kernel for the delta function approximation. `c` in this code may be a small positive number.

**Comment by Tony-Y** [comment]
@guolinke BTW, why does `b` exist?

Why not as the following?
```c++
return w * std::exp(-x * x / (2.0 * c * c)) * a / (c * std::sqrt(2 * pi));
```

**Comment by henry0312** [comment]
I'm sorry I don't have time to check the issue and remember the reason.

**Comment by henry0312** [comment]
Oh, I just rememberd!

> BTW, why does b exist?

b is the position of the center of the peak, so I wanted to clarify the shape of gaussian function.

cf. https://en.wikipedia.org/wiki/Gaussian_function

**Comment by henry0312** [comment]
When I have free time, I'll check this issue, sorry.

**Comment by henry0312** [comment]
cf. https://github.com/Microsoft/LightGBM/pull/178#issuecomment-271280902

**Comment by Tony-Y** [comment]
@henry0312 I have some questions.

1. Gradients are the Heaviside step function, but are not approximated as a smooth function in contrast of hessians that are the Kronecker delta function. Can this inconsistency be disregarded?

2. The small number `c` depends on both the variable `y` (score) and the constant `t` (label). Is it reasonable that `c` depends on `y`?

3. Can the global number `c` be determined from a statistical value of whole labels?

**Comment by Tony-Y** [comment]
If `c` must depend on `y`, I think  it should be a function of `x=|y-t|` in order to preserve the symmetry. 

For example,
```c++
const double diff = y - t;
const double x = std::fabs(diff);
const double c = std::max(x + std::fabs(t)) * eta, 1.0e-10);
```

**Comment by henry0312** [comment]
I have checked this issue and your questions and organised my answers, but it will take some time to write them in English (because of my poor English...).

**Comment by Tony-Y** [comment]
I found the objective function that has hessians in any form. A brief description is shown below.

We introduce the label weighting function `W_i(y)`, and define the objective function `L_t` as a weighted mean of loss function `l(y,y^)`. 
![obj_func](https://user-images.githubusercontent.com/11532812/31584835-8cb7a69c-b1f1-11e7-8ac2-79862d33d726.png)
The Taylor expansion of` L_t` is given as:
![taylor_exp](https://user-images.githubusercontent.com/11532812/31584874-164aac92-b1f2-11e7-87f0-8db70034deab.png)
![gradient](https://user-images.githubusercontent.com/11532812/31584911-be8c7c50-b1f2-11e7-9964-e12f07abe008.png)
![hessian](https://user-images.githubusercontent.com/11532812/31584923-ff7aa26e-b1f2-11e7-8fab-a792bd3cafb2.png)

If `l(y,y^) = |y - y^|`, we obtain:
![gradient_mae](https://user-images.githubusercontent.com/11532812/31585137-5cac3084-b1f7-11e7-9f27-e7ff48e86469.png)
![hessian_mae](https://user-images.githubusercontent.com/11532812/31585098-9e68574c-b1f6-11e7-8e50-6ef6e3cc8696.png)
Hence, we can have hessians in any form.

I hope users and developers have any idea.

**Comment by Tony-Y** [comment]
I take some notes on approximations to the delta or step function.

1. Logistic Approximation to Step Function
The logistic function f(x) is a smooth approximation to the step function.
https://en.m.wikipedia.org/wiki/Heaviside_step_function#Analytic_approximations
The derivative of f(x) is df/dx = f(x)f(1-x) = 𝞓(x), and an approximation to the delta function.
https://en.m.wikipedia.org/wiki/Logistic_function#Derivative
The weighting function can be defined as:
W_i(y) = 𝞓((y - y_i)/T)
where T is the temperature parameter.
In physics, f(x) is known as [the Fermi–Dirac distribution](https://en.wikipedia.org/wiki/Fermi–Dirac_statistics#Fermi.E2.80.93Dirac_distribution).

2. Methfessel-Paxton Approximation to Step Function
http://www.hector.ac.uk/cse/distributedcse/reports/conquest/conquest/node6.html
The N-th order approximation to the delta function:
   <img width="216" alt="mp-delta" src="https://user-images.githubusercontent.com/11532812/31595752-d8192bec-b278-11e7-91cf-c45bdd6441a7.png">
where A_n are suitable coefficients, and H_n are [Hermite polynomials](https://en.wikipedia.org/wiki/Hermite_polynomials).
The N-th order approximation to the step function:
   <img width="341" alt="mp-step" src="https://user-images.githubusercontent.com/11532812/31595894-af4f6978-b279-11e7-9dc9-90d46390e8fa.png">
   <img width="185" alt="mp-s0" src="https://user-images.githubusercontent.com/11532812/31595895-b2b22e16-b279-11e7-9535-13ea09ca6300.png">

3. Marzari-Vanderbilt Approximation to Delta Function
 https://arxiv.org/pdf/cond-mat/9903147.pdf
The approximation to the delta function:
    <img width="361" alt="mv-delta" src="https://user-images.githubusercontent.com/11532812/31596267-51392cae-b27c-11e7-9bdd-d9adfe4bb54d.png">
The approximation to the step function:
    <img width="353" alt="mv-step" src="https://user-images.githubusercontent.com/11532812/31596278-62b43b0e-b27c-11e7-93a7-7ca548d6fb08.png">

**Comment by Tony-Y** [comment]
Please see:
https://xgboost.readthedocs.io/en/latest/model.html

Suppose that each label y_i is not exactly equal to the prediction y_i^ for all time.
If the loss function is the mean absolute error, gradients are +1 or -1, and hessians are all zero.
If the L2 regularization parameter 𝝺 is non-zero, we obtain:
w_j* = - G_j/𝝺
obj_j* = -1/2 ∑ (G_j)^2/𝝺 + 𝛄T
On condition that the first assumption is reasonably valid, the algorithm probably works well.

**Comment by Tony-Y** [comment]
@henry0312 The above investigation would suggest that the approximate hessian you introduced is like a variable 𝝺. However, I can't get the exact mathematical expression yet.

What do you think about my suggestion?

**Comment by Tony-Y** [comment]
I conclude that Henry0312's approach is a special case of the following objective:
![objective](https://user-images.githubusercontent.com/11532812/31649122-4b5b0098-b34c-11e7-8a93-ff104ee2dcff.png)

![obj_t](https://user-images.githubusercontent.com/11532812/31649157-6d804534-b34c-11e7-867d-c8ec5dd90b57.png)

![omega](https://user-images.githubusercontent.com/11532812/31649171-7a8d0fa0-b34c-11e7-82e6-b30077c69739.png)

![lambda_j](https://user-images.githubusercontent.com/11532812/31649185-89880636-b34c-11e7-9a81-9f8c1a950a5c.png)
where
![index_set](https://user-images.githubusercontent.com/11532812/31649194-94947d34-b34c-11e7-8f41-cb23a88d5558.png)

When the loss function is the absolute error, we have to make it reasonably valid that each label y_i is not exactly equal to the prediction y_i^ for all time. In order to do so, the L2 weight function 𝛌 should be defined as a monatomic decreasing function of the absolute value of the difference between the label and the prediction:  
![lambda_func](https://user-images.githubusercontent.com/11532812/31649203-a1876ce0-b34c-11e7-8bbc-9797fb4d0ca2.png)
For example:
![phi](https://user-images.githubusercontent.com/11532812/31649214-ab649558-b34c-11e7-84d7-2ab4654c2337.png)

The L2 weight function 𝜆 resembles a hessian when the loss function is the absolute error. However,  𝜆 does not come from the approximation to the delta function.

@henry0312 If you agree with the above, please review and test your code on the basis of the objective with variable L2 weights.

**Comment by Tony-Y** [comment]
If we would like to remove the assumption that y_i ≠ y^_i when the loss function is the absolute error, we must ensure that
w_j* = 0
where i ∈ I_j and y_i = y^_i, because the sum of hessians H_j is infinity.

<img width="132" alt="w_j" src="https://user-images.githubusercontent.com/11532812/31701270-24dacbe2-b40a-11e7-86df-08388bddb158.png">
(This figure was taken from https://xgboost.readthedocs.io/en/latest/model.html)

**Comment by Tony-Y** [comment]
@guolinke I found two necessary conditions for using the absolute error as the loss function:

1. Use of the L2 regularization
2. Each label is not exactly equal to the corresponding prediction for all time

`Condition 1` is subject to the objective function, but `Condition 2` is subject to splitting a leaf.
In numerical implementation, when splitting a leaf  at step t, the following inequality must be satisfied:
![condition_2](https://user-images.githubusercontent.com/11532812/31715974-fc58d4de-b440-11e7-8e8e-3081eeba676f.png)
where 𝛆 > (machine epsilon).

Can you code this algorithm in the related source codes?

**Comment by henry0312** [comment]
Please wait until next week.
I'm too busy to work for LightGBM.

2017年10月18日(水) 20:54 Tony-Y <notifications@github.com>:

**Comment by gugatr0n1c** [comment]
Couple moths ago I run some tests on this. Using gradients {-1, 1} never worked good enough for me.
I got best results for L1 task with this:

def myObj(preds, train_data):
                        labels = train_data.get_label()
                        grad = np.sign(preds - labels) * np.abs(preds - labels)**0.5
                        hess = [1.] * len(preds)
                        return grad, hess
                    
Because gradients only with elements from {-1, 1} is "too stupid" for this algorithm.
The part: "np.abs(preds - labels)**0.5" helping algorithm to convergate but it is not so agressive as for L2. You can also run smaller exponent like: np.abs(preds - labels)**0.25

But this is only practical observation.

**Comment by Tony-Y** [comment]
@guolinke  BTW, do you really need the objective with the L1 loss? Actually, I don't need it.

**Comment by guolinke** [comment]
@Tony-Y I always use the solution like the @gugatr0n1c provided. 
And using some label conversions in regression is equivalent to change objective function.

**Comment by gugatr0n1c** [comment]
Btw when I run some stacking (multiple models with different hyper parameters) I always add some model  with objective like above. Even for L2 (MSE) task. The reason is: this form of gradients converge to different solution and brings new variability from data. It is like "free lunch". Almost in all datasets are some outliers pushing L2 norm to not so optimal solution (known drawback of L2).
But this is little bit offtopic.

**Comment by Tony-Y** [comment]
@guolinke Thank you for your reply. I had focused on mathematical aspect too much. I leave from this thread.

@gugatr0n1c  Thank you for letting me know your use case.

**Comment by henry0312** [comment]
I'll reply tonigt.

**Comment by Laurae2** [comment]
@henry0312 Any news?

**Comment by guolinke** [comment]
please try the one solution in #1199
